# EDA on Modbus Network Dataset

## Data Cleaning

#### Importing pandas, numpy and reading the dataset as Pandas Dataframe :

In [7]:
import pandas as pd
import numpy as np

In [8]:
df = pd.read_csv('Modbus_Data_EXP.csv')

In [9]:
df.head()

TopicNo  DeviceID          runId  count_of_frames_received_by_dstport  \
0        6  deviceID  1684399950:30                                 7927   
1        6  deviceID  1684399950:30                                 7928   
2        6  deviceID  1684399950:30                                 7177   
3        6  deviceID  1684399950:30                                 7177   
4        6  deviceID  1684399999:30                                  390   

   segment_max_when_push_flag_set  ttl_max  ttl_min  \
0                             260      126      126   
1                               9       64       64   
2                             260      126      126   
3                               9       60       60   
4                              36      128      128   

   count_when_reset_and_ack_flag_set  count_when_fin_or_syn_flag_set  \
0                                  0                               0   
1                                  0                               0   
2                                  0                               0   
3                                  0                               0   
4                                  0                               0   

   count_when_fin_and_syn_flag_set  ...  avg_MODbus_len  Unnamed: 93  \
0                                0  ...             104          NaN   
1                                0  ...               3          NaN   
2                                0  ...              78          NaN   
3                                0  ...               3          NaN   
4                                0  ...               6          NaN   

   Unnamed: 94  Unnamed: 95  Unnamed: 96  Unnamed: 97  count_of_read_op.1  \
0          NaN         7927            0          NaN            0.933014   
1          NaN         7928            0          NaN            0.932896   
2          NaN         7178            0          NaN            0.673865   
3          NaN         7177            0          NaN            0.673819   
4          NaN          390            0          NaN            0.789744   

   count_of_write_op.1  count_of_other_frame.1  count_of_diagnostic_op.1  
0             0.066986                     0.0                       0.0  
1             0.067104                     0.0                       0.0  
2             0.326135                     0.0                       0.0  
3             0.326181                     0.0                       0.0  
4             0.210256                     0.0                       0.0  

[5 rows x 102 columns]

#### Collecting the very basic info of the dataset :

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 311496 entries, 0 to 311495
Columns: 102 entries, TopicNo to count_of_diagnostic_op.1
dtypes: float64(7), int64(83), object(12)
memory usage: 242.4+ MB


The dataframe contains 311496 rows and 102 columns.
7 columns with float dtype
83 columns with int dtype
12 columns with object dtype

In [11]:
df.describe()

TopicNo  count_of_frames_received_by_dstport  \
count  311496.0                        311496.000000   
mean        6.0                           648.919453   
std         0.0                          4443.380881   
min         6.0                             0.000000   
25%         6.0                             1.000000   
50%         6.0                             1.000000   
75%         6.0                           111.000000   
max         6.0                        376139.000000   

       segment_max_when_push_flag_set        ttl_max        ttl_min  \
count                   311496.000000  311496.000000  311496.000000   
mean                        62.185675     100.233127     100.233127   
std                        102.916964      42.924580      42.924580   
min                          8.000000      60.000000      60.000000   
25%                          9.000000      62.000000      62.000000   
50%                          9.000000     126.000000     126.000000   
75%                         36.000000     128.000000     128.000000   
max                        422.000000     255.000000     255.000000   

       count_when_reset_and_ack_flag_set  count_when_fin_or_syn_flag_set  \
count                           311496.0                        311496.0   
mean                                 0.0                             0.0   
std                                  0.0                             0.0   
min                                  0.0                             0.0   
25%                                  0.0                             0.0   
50%                                  0.0                             0.0   
75%                                  0.0                             0.0   
max                                  0.0                             0.0   

       count_when_fin_and_syn_flag_set  count_when_fin_flag_set  \
count                         311496.0                 311496.0   
mean                               0.0                      0.0   
std                                0.0                      0.0   
min                                0.0                      0.0   
25%                                0.0                      0.0   
50%                                0.0                      0.0   
75%                                0.0                      0.0   
max                                0.0                      0.0   

       count_when_ack_flag_set  ...  avg_MODbus_len  Unnamed: 93  Unnamed: 94  \
count            311496.000000  ...   311496.000000          0.0          0.0   
mean                648.940738  ...       11.465203          NaN          NaN   
std                4443.400680  ...       21.335075          NaN          NaN   
min                   1.000000  ...        2.000000          NaN          NaN   
25%                   1.000000  ...        3.000000          NaN          NaN   
50%                   1.000000  ...        3.000000          NaN          NaN   
75%                 111.000000  ...        6.000000          NaN          NaN   
max              376139.000000  ...      254.000000          NaN          NaN   

         Unnamed: 95  Unnamed: 96  Unnamed: 97  count_of_read_op.1  \
count  311496.000000     311496.0          0.0         6390.000000   
mean      648.940738          0.0          NaN            0.285940   
std      4443.400680          0.0          NaN            0.376852   
min         1.000000          0.0          NaN            0.000000   
25%         1.000000          0.0          NaN            0.000000   
50%         1.000000          0.0          NaN            0.000000   
75%       111.000000          0.0          NaN            0.690027   
max    376139.000000          0.0          NaN            1.000000   

       count_of_write_op.1  count_of_other_frame.1  count_of_diagnostic_op.1  
count          6390.000000             6390.000000               6390.000000  
mean              0.091368                0.490141     

In [12]:
for col in df:
    print(col, df[col].dtype)

TopicNo int64
DeviceID object
runId object
count_of_frames_received_by_dstport int64
segment_max_when_push_flag_set int64
ttl_max int64
ttl_min int64
count_when_reset_and_ack_flag_set int64
count_when_fin_or_syn_flag_set int64
count_when_fin_and_syn_flag_set int64
count_when_fin_flag_set int64
count_when_ack_flag_set int64
count_when_syn_flag_set int64
count_when_reset_flag_set int64
count_when_push_flag_set int64
count_when_res_flag_set int64
count_when_ns_flag_set int64
count_when_cwr_flag_set int64
count_when_ecn_flag_set int64
count_when_urg_flag_set int64
count_when_all_flag_not_set int64
count_of_retransmitted_frames int64
count_of_duplicate_frames int64
count_of_BadTcp_frames int64
count_of_distinct_ttl int64
count_of_frames_sent int64
avg_tcp_length int64
avg_ip_length int64
avg_frame_length int64
avg_tcp_hdr_length int64
avg_ip_hdr_length int64
tcp_srcport int64
tcp_dstport int64
ip_src object
ip_dst object
ip_src_pair object
ip_dst_pair object
flowindicator int64
windowstartt

#### Checking the null values :

In [13]:
for col in df:
    print(col, df[col].isnull().sum())

TopicNo 0
DeviceID 0
runId 0
count_of_frames_received_by_dstport 0
segment_max_when_push_flag_set 0
ttl_max 0
ttl_min 0
count_when_reset_and_ack_flag_set 0
count_when_fin_or_syn_flag_set 0
count_when_fin_and_syn_flag_set 0
count_when_fin_flag_set 0
count_when_ack_flag_set 0
count_when_syn_flag_set 0
count_when_reset_flag_set 0
count_when_push_flag_set 0
count_when_res_flag_set 0
count_when_ns_flag_set 0
count_when_cwr_flag_set 0
count_when_ecn_flag_set 0
count_when_urg_flag_set 0
count_when_all_flag_not_set 0
count_of_retransmitted_frames 0
count_of_duplicate_frames 0
count_of_BadTcp_frames 0
count_of_distinct_ttl 0
count_of_frames_sent 0
avg_tcp_length 0
avg_ip_length 0
avg_frame_length 0
avg_tcp_hdr_length 0
avg_ip_hdr_length 0
tcp_srcport 0
tcp_dstport 0
ip_src 0
ip_dst 0
ip_src_pair 0
ip_dst_pair 0
flowindicator 0
windowstarttime 0
avg_eth_len 0
avg_ip_dsfield 0
count_of_ip_dsfield_ecn_is_3 0
count_when_ip_id_is_zero 0
count_of_ip_flags_df_is_set 0
count_of_ip_flags_df_is_reset 0
c

The Dataframe only consists prominent null values in the last 9 columns, in which 3 of them have all of their values as null.

#### Checking the unique values :

In [14]:
for col in df:
    print(col, df[col].nunique())

TopicNo 1
DeviceID 1
runId 6550
count_of_frames_received_by_dstport 10625
segment_max_when_push_flag_set 85
ttl_max 6
ttl_min 6
count_when_reset_and_ack_flag_set 1
count_when_fin_or_syn_flag_set 1
count_when_fin_and_syn_flag_set 1
count_when_fin_flag_set 1
count_when_ack_flag_set 10629
count_when_syn_flag_set 1
count_when_reset_flag_set 1
count_when_push_flag_set 10629
count_when_res_flag_set 1
count_when_ns_flag_set 1
count_when_cwr_flag_set 1
count_when_ecn_flag_set 1
count_when_urg_flag_set 1
count_when_all_flag_not_set 1
count_of_retransmitted_frames 8
count_of_duplicate_frames 1
count_of_BadTcp_frames 1
count_of_distinct_ttl 3
count_of_frames_sent 10629
avg_tcp_length 197
avg_ip_length 197
avg_frame_length 197
avg_tcp_hdr_length 12
avg_ip_hdr_length 11
tcp_srcport 71
tcp_dstport 90
ip_src 23
ip_dst 24
ip_src_pair 5
ip_dst_pair 19
flowindicator 31
windowstarttime 1
avg_eth_len 1
avg_ip_dsfield 1
count_of_ip_dsfield_ecn_is_3 1
count_when_ip_id_is_zero 18
count_of_ip_flags_df_is_set 

#### Dealing with null values by deletion :

In [15]:
for col in df.columns:
    if 'Unnamed' in col:
        if  df[col].isna().sum() == df.shape[0]:
            df = df.drop(columns=[col])

Deleting the Unnamed columns with complete column as NULL.

In [16]:
for col in df:
    print(col, df[col].isnull().sum())

TopicNo 0
DeviceID 0
runId 0
count_of_frames_received_by_dstport 0
segment_max_when_push_flag_set 0
ttl_max 0
ttl_min 0
count_when_reset_and_ack_flag_set 0
count_when_fin_or_syn_flag_set 0
count_when_fin_and_syn_flag_set 0
count_when_fin_flag_set 0
count_when_ack_flag_set 0
count_when_syn_flag_set 0
count_when_reset_flag_set 0
count_when_push_flag_set 0
count_when_res_flag_set 0
count_when_ns_flag_set 0
count_when_cwr_flag_set 0
count_when_ecn_flag_set 0
count_when_urg_flag_set 0
count_when_all_flag_not_set 0
count_of_retransmitted_frames 0
count_of_duplicate_frames 0
count_of_BadTcp_frames 0
count_of_distinct_ttl 0
count_of_frames_sent 0
avg_tcp_length 0
avg_ip_length 0
avg_frame_length 0
avg_tcp_hdr_length 0
avg_ip_hdr_length 0
tcp_srcport 0
tcp_dstport 0
ip_src 0
ip_dst 0
ip_src_pair 0
ip_dst_pair 0
flowindicator 0
windowstarttime 0
avg_eth_len 0
avg_ip_dsfield 0
count_of_ip_dsfield_ecn_is_3 0
count_when_ip_id_is_zero 0
count_of_ip_flags_df_is_set 0
count_of_ip_flags_df_is_reset 0
c

# Dealing null values by Mean and Mode :

In [17]:
mode_value = df['count_of_write_op.1'].mean()  
df['count_of_write_op.1'].fillna(mode_value, inplace=True) 

mode_value = df['count_of_read_op.1'].mean() 
df['count_of_read_op.1'].fillna(mode_value, inplace=True) 

These columns having null values replaced by mean of the column values because most of them were in between 0 and 1 and can be significantly important to the dataset.

In [18]:
mean_value = df['count_of_other_frame.1'].mode().iloc[0] 
df['count_of_other_frame.1'].fillna(mean_value, inplace=True)  

mean_value = df['count_of_diagnostic_op.1'].mode().iloc[0]  
df['count_of_diagnostic_op.1'].fillna(mean_value, inplace=True)  

These columns having null values replaced by mode of the column values because most of them were either 0 or 1 and can be significantly important to the dataset.

In [19]:
for col in df:
    print(col, df[col].isnull().sum())

TopicNo 0
DeviceID 0
runId 0
count_of_frames_received_by_dstport 0
segment_max_when_push_flag_set 0
ttl_max 0
ttl_min 0
count_when_reset_and_ack_flag_set 0
count_when_fin_or_syn_flag_set 0
count_when_fin_and_syn_flag_set 0
count_when_fin_flag_set 0
count_when_ack_flag_set 0
count_when_syn_flag_set 0
count_when_reset_flag_set 0
count_when_push_flag_set 0
count_when_res_flag_set 0
count_when_ns_flag_set 0
count_when_cwr_flag_set 0
count_when_ecn_flag_set 0
count_when_urg_flag_set 0
count_when_all_flag_not_set 0
count_of_retransmitted_frames 0
count_of_duplicate_frames 0
count_of_BadTcp_frames 0
count_of_distinct_ttl 0
count_of_frames_sent 0
avg_tcp_length 0
avg_ip_length 0
avg_frame_length 0
avg_tcp_hdr_length 0
avg_ip_hdr_length 0
tcp_srcport 0
tcp_dstport 0
ip_src 0
ip_dst 0
ip_src_pair 0
ip_dst_pair 0
flowindicator 0
windowstarttime 0
avg_eth_len 0
avg_ip_dsfield 0
count_of_ip_dsfield_ecn_is_3 0
count_when_ip_id_is_zero 0
count_of_ip_flags_df_is_set 0
count_of_ip_flags_df_is_reset 0
c

No Null values left in the Dataset.

#### Renaming the Unnamed columns :

In [20]:
unnamed_columns = [col for col in df.columns if 'Unnamed' in col]
new_column_names = ['NewName1', 'NewName2']
df.columns = [new_column_names.pop(0) if col in unnamed_columns else col for col in df.columns]

In [21]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 311496 entries, 0 to 311495
Data columns (total 99 columns):
 #   Column                                 Non-Null Count   Dtype  
---  ------                                 --------------   -----  
 0   TopicNo                                311496 non-null  int64  
 1   DeviceID                               311496 non-null  object 
 2   runId                                  311496 non-null  object 
 3   count_of_frames_received_by_dstport    311496 non-null  int64  
 4   segment_max_when_push_flag_set         311496 non-null  int64  
 5   ttl_max                                311496 non-null  int64  
 6   ttl_min                                311496 non-null  int64  
 7   count_when_reset_and_ack_flag_set      311496 non-null  int64  
 8   count_when_fin_or_syn_flag_set         311496 non-null  int64  
 9   count_when_fin_and_syn_flag_set        311496 non-null  int64  
 10  count_when_fin_flag_set                311496 non-null  

The dataframe has been cleaned and the final info can be analysed as above.

#### Saving the cleaned df into a csv :

In [22]:
df.to_csv('Cleaned_Modbus_data.csv')

Data Cleaning Done !